cpp/src/arrow/compute/kernels/vector_selection.cc

```diff
@@ -841,6 +907,222 @@ Status PrimitiveFilter(KernelContext* ctx, const ExecSpan& batch, ExecResult* ou
+                WriteNotNull(value_index, accumulated_run_length);
+              }
+              if (!is_valid) {
+                accumulated_run_length += run_length;
+                bit_util::ClearBit(out_is_valid_, out_position_);
+                WriteNull(accumulated_run_length);
+              }
+            });
+      }
+    } else {  // values_is_valid_ exists. Input may have nulls
+      if (filter_is_valid_ == NULLPTR) {
+        rle_util::VisitMergedRuns(
+            values_, filter_,
+            [&](int64_t run_length, int64_t value_index, int64_t filter_index) {
+              if (bit_util::GetBit(filter_data_, filter_index)) {
+                accumulated_run_length += run_length;
+                WriteMaybeNull(value_index, accumulated_run_length);
+              }
+            });
+      } else if (null_selection_ == FilterOptions::DROP) {
+        rle_util::VisitMergedRuns(
+            values_, filter_,
+            [&](int64_t run_length, int64_t value_index, int64_t filter_index) {
+              if (bit_util::GetBit(filter_is_valid_, filter_index) &&
+                  bit_util::GetBit(filter_data_, filter_index)) {
+                accumulated_run_length += run_length;
+                WriteMaybeNull(value_index, accumulated_run_length);
+              }
+            });
+      } else {  // null_selection == FilterOptions::EMIT_NULL
+        rle_util::VisitMergedRuns(
+            values_, filter_,
+            [&](int64_t run_length, int64_t value_index, int64_t filter_index) {
+              const bool is_valid = bit_util::GetBit(filter_is_valid_, filter_index);
+              if (is_valid && bit_util::GetBit(filter_data_, filter_index)) {
+                accumulated_run_length += run_length;
+                WriteMaybeNull(value_index, accumulated_run_length);
+              }
+              if (!is_valid) {
+                accumulated_run_length += run_length;
+                bit_util::ClearBit(out_is_valid_, out_position_);
+                WriteNull(accumulated_run_length);
+              }
+            });
+      }
+    }
+    out_logical_length_ = accumulated_run_length;
+  }
+
+  // Write the next out_position given the selected in_position for the input
+  // data and advance out_position
+  void WriteValue(int64_t in_position, int64_t run_length) {
+    out_run_length_[out_position_] = run_length;
+    out_data_[out_position_++] = values_data_[in_position];
+  }
+
+  void WriteNull(int64_t run_length) {
+    // Zero the memory
+    out_run_length_[out_position_] = run_length;
+    out_data_[out_position_++] = T{};
+  }
+
+ private:
+  const ArraySpan& values_;
+  const uint8_t* values_is_valid_;
+  const T* values_data_;
+  const ArraySpan& filter_;
+  const uint8_t* filter_is_valid_;
+  const uint8_t* filter_data_;
+  FilterOptions::NullSelectionBehavior null_selection_;
+  uint8_t* out_is_valid_;
+  int64_t* out_run_length_;
+  T* out_data_;
+  int64_t& out_logical_length_;
```

Comment:
I personally don't have a strong opinion about this but I think according to the style guide this should be a pointer (since it's an output).